### ISSUE TITLE
[Question] Consuming buffer list from nvim API

### ISSUE THREAD

**Issue by mackeybj** [Shougo/denite.nvim#165]
I am working on adding support for the buffer_tab source using [tabpagebuffer.vim](https://github.com/Shougo/tabpagebuffer.vim). I have copied the implementation of buffer.py in denite sources and named it buffer_tab.py and changed the relevant strings.

When I [gather candidates](https://github.com/Shougo/denite.nvim/blob/master/rplugin/python3/denite/source/buffer.py#L50) and try to compare the tab indexes from tabpagebuffer.vim using `self.vim.eval('t:tabpagebuffer')` with the indexes from `self.vim.buffers`, I am unable to iterate over the [Buffers](https://github.com/neovim/python-client/blob/master/neovim/api/nvim.py#L351) iterator properly.

Whenever I try to access the `number` property of buffer while iterating over the Buffers object, I get an error that Buffers has no property `number`. The same code from an iPython interpreter attached to the same neovim instance is able to run properly and create a list of buffer, rather than only using the Buffers iterator.

I have tried extending the existing [`candidates`](https://github.com/Shougo/denite.nvim/blob/master/rplugin/python3/denite/source/buffer.py#L51) list generator with a conditional, accessing from within a separate list generator, casting the generator to a list with both `list(self.vim.buffers)` and `[self.vim.buffers]` and then iterating over the elements, as well as using `self.vim.api.list_bufs` but still get the Buffers iterator object. Even making a new variable with `shortlist = list(self.vim.buffers)` only returns a Buffers iterator.

I am quite confident this behaviour is caused by either denite.nvim or python-client API but I'm not sure which or how to debug this strange behaviour. Any pointers in the right direction would be great!

TL;DR iPython and :terminal can consume Buffers iterator. Same python run from rplugin does not

**Comment by Shougo** [comment]
Hm.  I want to reproduce the problem.
Please upload the code.

**Comment by mackeybj** [comment]
This is the whole [buffer_tab.py](https://github.com/mackeybj/denite.nvim/blob/master/rplugin/python3/denite/source/buffer_tab.py#L51) file.

The relevant function is below where I have used a simple method to show the problem

```
def gather_candidates(self, context):
        shortlist = []
        for buffer in self.vim.buffers:
            if str(buffer.number) in self.vim.eval('t:tabpagebuffer').keys():
                shortlist.append(buffer)
        print(shortlist)

        candidates = [
            self._convert(ba) for ba in [
                bufattr for bufattr in [
                    self._get_attributes(context, buf)
                    for buf in shortlist
                ] if not self._is_excluded(context, bufattr)
            ]
        ]

        return sorted(candidates, key=(
            lambda x:
            maxsize if context['__caller_bufnr'] == x['bufnr']
            else -maxsize if context['__alter_bufnr'] == x['bufnr']
            else x['timestamp']))
```

**Comment by Shougo** [comment]
OK.  I will try it later.

**Comment by Shougo** [comment]
`print(shortlist)` does not work in the remote plugin context.

**Comment by mackeybj** [comment]
My apologies. I added the code back in quickly after removing it. I had made use of error from util for debug output

**Comment by Shougo** [comment]
I have tested your source and it works well....
So I don't think it is problem.

**Comment by mackeybj** [comment]
You're entirely correct. It now works as expected for me as well. I have no idea why I was getting this behaviour yesterday even after cleaning plugins and restarting all neovim processes.

Sorry for wasting your time and thank you for your help
